File: src/services/auth-service/src/services/audit-emitter.ts
/**
 * Emits RBAC governance events to the platform audit trail (Kafka `traverse.cpa.audit-events`),
 * which audit-service consumes into its immutable hash-chained store (Phase 3).
 *
 * Previously auth-service logged role/permission changes to the app log only, so
 * "who changed which role, when" was not in the tamper-evident audit trail. This
 * connects it, mirroring the .NET AuditEmitter event shape
 * ({ EventType, TimestampUtc, UserId, EntityType, EntityId, BeforeState?, AfterState? }).
 *
 * Best-effort by design: an audit emit must NEVER fail an RBAC operation. The
 * producer connects lazily, and every failure is logged, not thrown. If
 * KAFKA_BOOTSTRAP_SERVERS is unset the emitter is a no-op (local dev without Kafka).
 */

import { Kafka, Producer, logLevel } from 'kafkajs';
import logger from '../config/logger';

const TOPIC = 'traverse.cpa.audit-events';

class AuditEmitter {
  private producer: Producer | null = null;
  private connecting: Promise<void> | null = null;
  private readonly enabled: boolean;

  constructor() {
    const brokers = (process.env.KAFKA_BOOTSTRAP_SERVERS || '')
      .split(',')
      .map((b) => b.trim())
      .filter(Boolean);
    this.enabled = brokers.length > 0;
    if (!this.enabled) {
      logger.info('KAFKA_BOOTSTRAP_SERVERS not set — RBAC audit emit disabled');
      return;
    }
    const kafka = new Kafka({ clientId: 'auth-service', brokers, logLevel: logLevel.ERROR });
    this.producer = kafka.producer({ allowAutoTopicCreation: true });
  }

  private async ensureConnected(): Promise<void> {
    if (!this.producer) return;
    if (!this.connecting) {
      this.connecting = this.producer.connect().catch((err) => {
        this.connecting = null; // allow a later retry
        throw err;
      });
    }
    return this.connecting;
  }

  /**
   * Fire-and-forget audit emit. `actorUserId` is the admin performing the change;
   * `entityId` is the role name or user id affected.
   */
  emit(
    eventType: string,
    actorUserId: string | undefined,
    entityType: 'Role' | 'User' | 'RolePermissions',
    entityId: string,
    before?: unknown,
    after?: unknown
  ): void {
    if (!this.enabled || !this.producer) return;
    const payload = JSON.stringify({
      EventType: eventType,
      TimestampUtc: new Date().toISOString(),
      UserId: actorUserId ?? 'system',
      EntityType: entityType,
      EntityId: entityId,
      ...(before !== undefined ? { BeforeState: before } : {}),
      ...(after !== undefined ? { AfterState: after } : {}),
    });

    void this.ensureConnected()
      .then(() => this.producer!.send({ topic: TOPIC, messages: [{ value: payload }] }))
      .catch((err) => logger.warn(`RBAC audit emit failed for ${eventType}/${entityId}: ${err?.message ?? err}`));
  }

  async shutdown(): Promise<void> {
    if (this.producer && this.connecting) {
      try { await this.producer.disconnect(); } catch { /* ignore */ }
    }
  }
}

export const auditEmitter = new AuditEmitter();

File: src/services/display-service/Services/AuditEmitter.cs
using System.Text.Json;
using Confluent.Kafka;

namespace Traverse.DisplayService.Services;

/// <summary>
/// Emits governance audit events to the platform audit trail (Kafka topic <c>traverse.cpa.audit-events</c>), which
/// audit-service consumes into its immutable hash-chained store. Phase 5 — this connects the previously
/// disconnected display-events stream to the real audit log instead of inventing a third path.
///
/// Best-effort by design: an audit emit must never fail a display operation, so produce is fire-and-forget
/// and a Kafka outage is logged, not surfaced. The event shape mirrors AMS.AuditService.Models.AuditEvent
/// (the consumer deserializes case-insensitively and computes the hash chain itself).
/// </summary>
public sealed class AuditEmitter : IDisposable
{
    private readonly IProducer<Null, string>? _producer;
    private readonly ILogger<AuditEmitter> _logger;
    private const string Topic = "traverse.cpa.audit-events";

    public AuditEmitter(IConfiguration config, ILogger<AuditEmitter> logger)
    {
        _logger = logger;
        var bootstrap = config["Kafka:BootstrapServers"];
        if (string.IsNullOrWhiteSpace(bootstrap))
        {
            _logger.LogInformation("Kafka:BootstrapServers not configured — audit emit disabled");
            return;
        }
        try
        {
            _producer = new ProducerBuilder<Null, string>(new ProducerConfig
            {
                BootstrapServers = bootstrap,
                // Never let a slow/absent broker block the request thread.
                MessageTimeoutMs = 5000,
                Acks = Acks.Leader,
            }).Build();
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Could not build Kafka producer — audit emit disabled");
        }
    }

    public void Emit(string eventType, string userId, Guid entityId, string entityName, object? after = null)
    {
        if (_producer is null) return;
        try
        {
            var payload = JsonSerializer.Serialize(new
            {
                EventType = eventType,
                TimestampUtc = DateTimeOffset.UtcNow,
                UserId = userId,
                EntityType = "Display",
                EntityId = entityId.ToString(),
                AfterState = after ?? new { name = entityName },
            });
            _producer.Produce(Topic, new Message<Null, string> { Value = payload }, report =>
            {
                if (report.Error.IsError)
                    _logger.LogWarning("Audit emit failed: {Reason}", report.Error.Reason);
            });
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Audit emit threw for {EventType}/{EntityId}", eventType, entityId);
        }
    }

    public void Dispose() => _producer?.Dispose();
}
